🏗 Allow objstr()

alanorozco opened the pull request:
- [Adds `obj-str` dependency.](https://github.com/lukeed/obj-str)

- Lints to only allow literal object expressions: 
    - ✅ `objstr({'my-class': maybe})`
    - ❌ `objstr('bad')`
    - ❌ `objstr(bad)`
    - ❌ `objstr({...bad})`

jridgewell commented on build-system/babel-plugins/babel-plugin-optimize-objstr/index.js:
```diff
@@ -0,0 +1,69 @@
+    name: 'optimize-objstr',
+    visitor: {
+      CallExpression(path) {
+        const {name} = path.node.callee;
+        if (name !== 'objstr') {
+          return;
+        }
```
```suggestion
        const callee = path.get('callee');
        if (!callee.referencesImport('objstr', 'default')) {
          return;
        }
```

alanorozco replied:
This is so much better.

jridgewell commented on build-system/babel-plugins/babel-plugin-optimize-objstr/test/fixtures/transform-assertions/transform-mixed/output.js:
```diff
@@ -0,0 +1,22 @@
+"" + (keyIsValueId ? " " + "keyIsValueId" : "");
+"" + (false ? " " + 'quoted' : "");
+"" + (true ? " " + "unquoted" : "");
+"" + (probably ? " " + "identifierExpression" : "");
```
Shouldn't this be:
```suggestion
"" + (probably ? " " + identifierExpression : "");
```

alanorozco replied:
Good catch! Fixed using your suggestion below.

jridgewell commented on build-system/babel-plugins/babel-plugin-optimize-objstr/index.js:
```diff
@@ -0,0 +1,69 @@
+              t.binaryExpression(
+                '+',
+                t.stringLiteral(' '),
+                t.isIdentifier(prop.key)
```
```suggestion
                t.isIdentifier(prop.key) && !prop.computed
```

I believe this will fail:

```js
objStr({
  [key]: true
});
```

That should generate `true ? " " + key : ""`, right?

alanorozco replied:
Yep, fixed!

jridgewell commented on build-system/babel-plugins/babel-plugin-optimize-objstr/index.js:
```diff
@@ -0,0 +1,69 @@
+            }
+            const conditional = t.conditionalExpression(
+              prop.value,
+              t.binaryExpression(
```
Using `babel.template` could help clean this up?

```js
const conditional = template.expression.ast`${prop.value} ? " " + ${prop.key} : ""`;
```

alanorozco replied:
Did not know this existed. It's so much better too!

dvoytenko commented:
Concept and use of `objstr` is LGTM. The only question is where the babel plugin will be hosted.

alanorozco commented:
I've removed the Babel plugin from this change in favor of lukeed/obj-str#6. I've incorporated @jridgewell's suggestions for the transform in that PR, in addition to options `{strict, leadingSpace, referencesImport}`.
